src/server.rs

```diff
@@ -339,46 +350,55 @@ mod secure_server {
     }
 }
 
-pub struct Inner {
-    env: Arc<Environment>,
+struct ServerCore {
     server: *mut GrpcServer,
     bind_addrs: Vec<(String, u16)>,
     slots_per_cq: usize,
     shutdown: AtomicBool,
-    handlers: HashMap<&'static [u8], Handler>,
 }
 
-impl Inner {
-    /// Get the handler for the requested method path.
-    pub fn get_handler(&self, method: &[u8]) -> Option<&Handler> {
-        self.handlers.get(method)
+impl Drop for ServerCore {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpc_server_destroy(self.server) }
     }
 }
 
-impl Debug for Inner {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(f, "Server {:?}", self.bind_addrs)
-    }
+unsafe impl Send for ServerCore {}
+unsafe impl Sync for ServerCore {}
+
+pub type BoxHandler = Box<CloneableHandler>;
+
+#[derive(Clone)]
+pub struct RequestCallContext {
+    server: Arc<ServerCore>,
+    registry: Arc<HashMap<&'static [u8], BoxHandler>>,
 }
 
-impl Drop for Inner {
-    fn drop(&mut self) {
-        unsafe { grpc_sys::grpc_server_destroy(self.server) }
+impl RequestCallContext {
+    /// Users should gurantee the method is always called from the same thread.
```

Comment:
`guarantee`

One way to enforce this is to have the type change to a non-`Send` type after this function is called.

Reply:
I don't see any convenient way to do it. The communication between threads happen inside unsafe c codes, which doesn't have the concept of `Send`.